helix-core/src/transaction.rs

```diff
@@ -513,6 +542,44 @@ impl Transaction {
         Self::change(doc, selection.iter().map(f))
     }
 
+    pub fn change_by_selection_ignore_overlapping(
+        doc: &Rope,
+        selection: &Selection,
+        mut change_range: impl FnMut(&Range) -> (usize, usize),
+        mut create_tendril: impl FnMut(usize, usize) -> Option<Tendril>,
+    ) -> (Transaction, Selection) {
+        let mut last_selection_idx = None;
+        let mut new_primary_idx = None;
+        let mut ranges: SmallVec<[Range; 1]> = SmallVec::new();
+        let process_change = |change_start, change_end, (idx, range): (usize, &Range)| {
+            // update the primary idx
+            if idx == selection.primary_index() {
+                new_primary_idx = Some(idx);
+            } else if new_primary_idx.is_none() {
+                if idx > selection.primary_index() {
+                    new_primary_idx = last_selection_idx;
+                } else {
+                    last_selection_idx = Some(idx);
+                }
+            }
+            ranges.push(*range);
+            create_tendril(change_start, change_end)
+        };
+        let transaction = Self::change_ignore_overlapping(
+            doc,
+            selection.iter().enumerate().map(|range| {
+                let (change_start, change_end) = change_range(range.1);
+                (change_start, change_end, range)
+            }),
+            process_change,
+        );
+
+        (
+            transaction,
+            Selection::new(ranges, new_primary_idx.unwrap_or(0)),
```

Comment:
Both callers of change_by_selection_ignore_overlapping end up applying transaction to the selection and using that resulting selection (generate_transaction_from_completion_edit does that directly, generate_transaction_from_snippet applies changes to each range). Couldn't you just apply transaction to selection here and attach that resulting selection to the transaction?

Reply:
The selection is changed again in both cases and there is no way to remove the selection fron a transaction (and even if there were it would require an unwrap). 

Applying the selection to the transaction would also be a footgun/semantically incorret because the returned selection is not mapped trough changes yet (and we can't do that inside this function since mapping can cause ranges to be sorted/merged which would mess up the zip for snippets), instead this selection is identical to to inout aelection with only those ranges whose edits have been dropped removed.

Reply:
> and we can't do that inside this function since mapping can cause ranges to be sorted/merged which would mess up the zip for snippets

I was suggesting returning the mapped selection (sorry if I didn't use the right terminology and made it confusing). I think the sorting/merging aspect is what I missed.